code/assignTime.m

```diff
@@ -0,0 +1,281 @@
+packetsToKeep = setdiff(1:size(dataTable_original,1),packetsToRemove);
+dataTable = dataTable_original(packetsToKeep,:);
+
+%%
+% Identify gaps -- start with most obvious gaps, and then become more
+% refined
+
+% Change in sampling rate should start a new chunk; identify indices of
+% last packet of a chunk
+if ~isequal(length(unique(dataTable.samplerate)),1)
+    indices_changeFs = find(diff(dataTable.samplerate) ~=0);
+else
+    indices_changeFs = [];
+end
+maxFs = max(unique(dataTable.samplerate));
+
+% Timestamp difference from previous packet not 0 or 1 -- indicates gap or
+% other abnormality; these indices indicate the end of a continuous chunk
+indices_timestampFlagged = intersect(find(diff(dataTable.timestamp) ~= 0),find(diff(dataTable.timestamp) ~= 1));
+
+% Instances where dataTypeSequence doesn't iterate by 1; these indices
+% indicate the end of a continuous chunk
+indices_dataTypeSequenceFlagged = intersect(find(diff(dataTable.dataTypeSequence) ~= 1),find(diff(dataTable.dataTypeSequence) ~= -255));
+
+% Lastly, check systemTick to ensure there are no packets missing.
+% Determine delta time between systemTicks in adjacent packets.
+% Exclude first packet from those to test, because need to compare times from
+% 'previous' to 'current' packets; diff_systemTick written to index of
+% second packet associated with that calculation
+numPackets = size(dataTable,1);
+for iPacket = 2:numPackets
+    diff_systemTick(iPacket,1) = mod((dataTable.systemTick(iPacket) + (2^16)...
+        - dataTable.systemTick(iPacket - 1)), 2^16);
+end
+
+% Expected elapsed time for each packet, based on sample rate and number of
+% samples per packet; in units of systemTick (1e-4 seconds)
+expectedElapsed = dataTable.packetsizes .* (1./dataTable.samplerate) * 1e4;
+
+% If diff_systemTick and expectedElapsed differ by more than 10% of expectedElapsed,
+% flag as gap
+indices_systemTickFlagged = find (abs(expectedElapsed(2:end) - diff_systemTick(2:end)) > 0.1*expectedElapsed(2:end));
+
+% All packets flagged as end of continuous chunks
+allFlaggedIndices = unique([indices_changeFs; indices_timestampFlagged;...
+    indices_dataTypeSequenceFlagged; indices_systemTickFlagged],'sorted') ;
+
+%%
+disp('Chunking data')
+% Determine indices of packets which correspond to each data chunk
+if ~isempty(allFlaggedIndices)
+    counter = 1;
+    chunkIndices = cell(1,length(allFlaggedIndices) + 1);
+    for iChunk = 1:length(allFlaggedIndices)
+        if iChunk == 1
+            chunkIndices{counter} = (1:allFlaggedIndices(1));
+            currentStartIndex = 1;
+            counter = counter + 1;
+        else
+            chunkIndices{counter} = allFlaggedIndices(currentStartIndex) + 1:allFlaggedIndices(currentStartIndex + 1);
+            currentStartIndex = currentStartIndex + 1;
+            counter = counter + 1;
+        end
+        % Last chunk, finishing with last packet
+        chunkIndices{counter} = allFlaggedIndices(currentStartIndex) + 1:numPackets;
```

Comment:
A lot of unnecessary assignment into a cell location. For all but the last iteration of the loop, the data in that location is reassigned. There are different ways to change the loop to cut down on time. I have put one solution that changes things only a little bit below. I only tested it in one data set though: Session1568860301794/DeviceNPC700419H/RawDataTD.json. I think this was the longest recording I could find in the Box folders. 

I changed the loop to check if it is in the last iteration of the loop and filled the last two entries accordingly.  I saw a decrease in time from ~12 sec to 0.1 sec on the data set used above. I am using a iMac 2019 edition with MATLAB 2018.

```suggestion

if ~isempty(allFlaggedIndices)
    counter = 1;
    chunkIndices2 = cell(1,length(allFlaggedIndices) + 1);
    for iChunk = 1:length(allFlaggedIndices)
        if iChunk == 1
            chunkIndices2{counter} = (1:allFlaggedIndices(1));
            currentStartIndex = 1;
            counter = counter + 1;
        elseif iChunk == length(allFlaggedIndices)
            chunkIndices2{counter} = allFlaggedIndices(currentStartIndex) + 1:allFlaggedIndices(currentStartIndex + 1);
            chunkIndices2{counter+1} = allFlaggedIndices(currentStartIndex+1) + 1:numPackets;
        else
            chunkIndices2{counter} = allFlaggedIndices(currentStartIndex) + 1:allFlaggedIndices(currentStartIndex + 1);
            currentStartIndex = currentStartIndex + 1;
            counter = counter + 1;
        end
    end
else
    % No identified missing packets, all packets in one chunk
    chunkIndices{1} = 1:numPackets;
end
```

Reply:
Resolved with https://github.com/openmind-consortium/UCSF-rcs-data-analysis/commit/4ffcae05b8e8cf1516e4d07015b74ad651a38e75